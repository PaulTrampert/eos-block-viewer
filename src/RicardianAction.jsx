import React from 'react';
import PropTypes from 'prop-types';
import contractFactory from './ricardianContractFactory';
import styles from './RicardianAction.module.css';
import RicardianMeta from './RicardianMeta';
import Transaction from './propTypes/Transaction';

class RicardianAction extends React.Component {
  render = () => {
    const {
      transaction,
      actionIndex,
      abi
    } = this.props;
    try {
      const contract = contractFactory.create({
        abi,
        transaction,
        actionIndex,
        allowUnusedVariables: true
      });
      const meta = contract.getMetadata();
      const html = {
        __html: contract.getHtml()
      };

      return (
        <div className={styles.ricardian}>
          <RicardianMeta {...meta} />
          <div dangerouslySetInnerHTML={html} />
        </div>
      );
    } catch (e) {
      var action = transaction.actions[actionIndex];
      return (
        <div>
          <RicardianMeta title={action.account} summary={action.name} />
          <p>{JSON.stringify(action.data)}</p>
        </div>
      )
    }
  }
}

RicardianAction.propTypes = {
  transaction: Transaction,
  abi: PropTypes.object,
  actionIndex: PropTypes.number
};

export default RicardianAction;