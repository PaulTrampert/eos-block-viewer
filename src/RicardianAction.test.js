import React from 'react';
import {shallow} from 'enzyme'
import contractFactory from './ricardianContractFactory';
import RicardianAction from './RicardianAction';

jest.mock('./ricardianContractFactory');

describe('RicardianAction', () => {
  let subject;
  let transaction;
  let abi;
  let actionIndex;
  let contract;
  let meta;
  let html;

  beforeEach(() => {
    transaction = {
      actions: [
        {
          account: "Test",
          name: "action",
          data: {
            some: 'data'
          }
        }
      ]
    };
    abi = {}
    actionIndex = 0;
    meta = {
      title: 'The Everlasting Compact'
    }
    html = "<p>Hi there! I'm a contract!</p>";
    contract = {
      getMetadata: jest.fn().mockReturnValue(meta),
      getHtml: jest.fn().mockReturnValue(html)
    };
  })

  describe('render', () => {
    describe("when the contract gets rendered just fine", () => {
      beforeEach(() => {
        contractFactory.create.mockReturnValue(contract);
        subject = shallow(<RicardianAction transaction={transaction} abi={abi} actionIndex={actionIndex} />);
      });

      it("renders the contract meta and html", () => {
        expect(subject).toMatchSnapshot();
      });
    });

    describe('when there is an error rendering the contract', () => {
      beforeEach(() => {
        contractFactory.create.mockImplementation(() => {
          throw new Error("it broke!");
        });
        subject = shallow(<RicardianAction transaction={transaction} abi={abi} actionIndex={actionIndex} />);
      })

      it('renders an error view', () => {
        expect(subject).toMatchSnapshot();
      })
    });
  });
});