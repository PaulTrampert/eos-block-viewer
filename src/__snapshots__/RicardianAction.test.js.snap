// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RicardianAction render when the contract gets rendered just fine renders the contract meta and html 1`] = `
<div
  className="ricardian"
>
  <RicardianMeta
    title="The Everlasting Compact"
  />
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<p>Hi there! I'm a contract!</p>",
      }
    }
  />
</div>
`;

exports[`RicardianAction render when there is an error rendering the contract renders an error view 1`] = `
<div>
  <RicardianMeta
    summary="action"
    title="Test"
  />
  <p>
    {"some":"data"}
  </p>
</div>
`;
